lib/posix/pthread.c

```diff
@@ -377,28 +394,36 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
  */
 void pthread_exit(void *retval)
 {
-	k_spinlock_key_t cancel_key;
+	k_spinlock_key_t key;
 	struct posix_thread *self = to_posix_thread(pthread_self());
 	pthread_key_obj *key_obj;
 	pthread_thread_data *thread_spec_data;
 	sys_snode_t *node_l;
 
-	/* Make a thread as cancelable before exiting */
-	cancel_key = k_spin_lock(&self->cancel_lock);
-	if (self->cancel_state == PTHREAD_CANCEL_DISABLE) {
-		self->cancel_state = PTHREAD_CANCEL_ENABLE;
-	}
+	__ASSERT(self != NULL, "not a valid pthread (%p)", k_current_get());
```

Comment:
Hrm. Is pthread_t at least as large as size_t and unisigned? Otherwise, I think `to_posix_thread()` needs some better bounds checking to avoid possible wrap issues (imagine a 32-bit pthread_t on a  64-bit size_t system, it's possible that a large address difference will wrap around back into range).

Reply:
`pthread_t` is a `uint32_t` in Zephyr, so should be safe.

Reply:
that makes it unsafe -- on 64 bit targets, size_t is 64 bits, so you will get aliasing when converting from size_t to pthread_t. Probably need to fix pthread_self to bounds check the computed pointer before converting to an array index with the subtraction.

Reply:
True - for some reason I was thinking it was 32-bit. I'll add a bounds check in this rev